frontend/src/functions/authorization.ts

```diff
@@ -0,0 +1,90 @@
+        if (permissions[namespace][resource]) {
+            if (permissions[namespace][resource].includes(operation)) {
+                return true;
+            }
+        } else if (permissions[namespace]["*"].includes(operation)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+export function hasAnyPermissionInAnyNamespace(userStatus: UserStatus, resource: string): boolean {
+    const permissions = userStatus.permissions;
+    for (const namespace in permissions) {
+        if (permissions[namespace][resource]) {
+            if (permissions[namespace][resource].length > 0) {
+                return true;
+            }
+        } else if (permissions[namespace]["*"].length > 0) {
+            return true;
+        }
+    }
+    return false;
+}
+
+export function hasPermissionInAnyResource(userStatus: UserStatus, namespace: string, operation: "c" | "r" | "u" | "d" | "l"): boolean {
+    const permissions = userStatus.permissions;
+    for (const resource in permissions[namespace]) {
+        if (permissions[namespace][resource].includes(operation)) {
+            return true;
+        }
+    }
+    for (const resource in permissions["*"]) {
+        if (!permissions[namespace][resource] && permissions["*"][resource].includes(operation)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+export function allowedNamespaces(userStatus: UserStatus, operation: "c" | "r" | "u" | "d" | "l", resource: string): string[] {
+    const permissions = userStatus.permissions;
+    const namespaces: string[] = [];
+    for (const namespace in permissions) {
+        if (permissions[namespace][resource]) {
+            if (permissions[namespace][resource].includes(operation)) {
+                namespaces.push(namespace);
+            }
+        } else if (permissions[namespace]["*"].includes(operation)) {
+            namespaces.push(namespace);
+        }
+    }
+    return namespaces;
+}
+
+export function allowedResources(userStatus: UserStatus, operation: "c" | "r" | "u" | "d" | "l", namespace: string): string[] {
+    const permissions = userStatus.permissions;
+    const resources: string[] = [];
+    for (const resource in permissions[namespace]) {
+        if (permissions[namespace][resource].includes(operation)) {
+            resources.push(resource);
+        }
+    }
+    for (const resource in permissions["*"]) {
+        if (!permissions[namespace][resource] && permissions["*"][resource].includes(operation)) {
+            resources.push(resource);
+        }
+    }
+    return resources;
+}
+
+export function allowedOperations(userStatus: UserStatus, namespace: string, resource: string): string[] {
+    const permissions = userStatus.permissions;
+    const lookupNs = permissions[namespace] ? namespace : "*";
+    const lookupRes = permissions[lookupNs] && permissions[lookupNs][resource] ? resource : "*";
+    return permissions[lookupNs][lookupRes] || [];
+}
```

Comment:
Let me know what you think—some of these might be better than yours, though not all.

```suggestion
type Operation = "c" | "r" | "u" | "d" | "l";

function getEffectivePermissions(permissions: UserStatus['permissions'], namespace: string, resource: string): Operation[] {
    const nsPermissions = permissions[namespace] || permissions["*"] || {};
    const resPermissions = nsPermissions[resource] || nsPermissions["*"] || [];
    return resPermissions as Operation[];
}

export function hasPermission(userStatus: UserStatus, namespace: string, resource: string, operation: Operation): boolean {
    const resPermissions = getEffectivePermissions(userStatus.permissions, namespace, resource);
    return resPermissions.includes(operation);
}

export function hasPermissionInAnyNamespace(userStatus: UserStatus, resource: string, operation: Operation): boolean {
    const permissions = userStatus.permissions;
    return Object.keys(permissions).some((namespace) => {
        const resPermissions = getEffectivePermissions(permissions, namespace, resource);
        return resPermissions.includes(operation);
    });
}

export function hasAnyPermissionInAnyNamespace(userStatus: UserStatus, resource: string): boolean {
    const permissions = userStatus.permissions;
    return Object.keys(permissions).some((namespace) => {
        const resPermissions = getEffectivePermissions(permissions, namespace, resource);
        return resPermissions.length > 0;
    });
}

export function hasPermissionInAnyResource(userStatus: UserStatus, namespace: string, operation: Operation): boolean {
    const nsPermissions = userStatus.permissions[namespace] || userStatus.permissions["*"];
    if (!nsPermissions) return false;

    return Object.keys(nsPermissions).some((resource) => {
        const resPermissions = getEffectivePermissions(userStatus.permissions, namespace, resource);
        return resPermissions.includes(operation);
    });
}

export function allowedNamespaces(userStatus: UserStatus, operation: Operation, resource: string): string[] {
    const permissions = userStatus.permissions;
    const namespaces = new Set<string>();

    Object.keys(permissions).forEach((namespace) => {
        const resPermissions = getEffectivePermissions(permissions, namespace, resource);
        if (resPermissions.includes(operation)) {
            namespaces.add(namespace);
        }
    });

    return Array.from(namespaces);
}

export function allowedResources(userStatus: UserStatus, operation: Operation, namespace: string): string[] {
    const permissions = userStatus.permissions;
    const resources = new Set<string>();

    const nsPermissions = permissions[namespace] || {};
    const wildcardNsPermissions = permissions["*"] || {};

    const combinedPermissions = { ...wildcardNsPermissions, ...nsPermissions };

    Object.keys(combinedPermissions).forEach((resource) => {
        const resPermissions = getEffectivePermissions(permissions, namespace, resource);
        if (resPermissions.includes(operation)) {
            resources.add(resource);
        }
    });

    return Array.from(resources);
}

export function allowedOperations(userStatus: UserStatus, namespace: string, resource: string): Operation[] {
    return getEffectivePermissions(userStatus.permissions, namespace, resource);
}
```

Reply:
I agree some of these are cleaner code, however I see some bugs in some of the refactored ones you shared. To accept this refactor I would have to break my mind on these and analize if there are any edge cases I might have overlooked. Can we agree to disagree - I'll add the operation type, and some of the smaller refactors, but most of the functions will stay the same. If you see any fault in my logic or way I am using javascript please point out the specific case